src/Microsoft.VisualStudio.Threading/AsyncLazy`1.cs

```diff
@@ -234,6 +261,113 @@ public T GetValue(CancellationToken cancellationToken)
+    /// <para>If the value has already been produced and implements <see cref="IDisposable"/> or <see cref="IAsyncDisposable"/>, it will be disposed of.
+    /// If the value factory has already started but has not yet completed, its value will be disposed of when the value factory completes.</para>
+    /// <para>If prior calls to obtain the value are in flight when this method is called, those calls <em>may</em> complete and their callers may obtain the value, although <see cref="IDisposable.Dispose"/>
+    /// may have been or will soon be called on the value, leading those users to experience a <see cref="ObjectDisposedException"/>.</para>
+    /// <para>Note all conditions based on the value implementing <see cref="IDisposable"/> or <see cref="IAsyncDisposable"/> is based on the actual value, rather than the <typeparamref name="T"/> type argument.
+    /// This means that although <typeparamref name="T"/> may be <c>IFoo</c> (which does not implement <see cref="IDisposable"/>), the concrete type that implements <c>IFoo</c> may implement <see cref="IDisposable"/>
+    /// and thus be treated as a disposable object as described above.</para>
+    /// </remarks>
+    public Task DisposeValueAsync()
+    {
+        Task<T>? localValueTask = null;
+        IDisposable? localValue = default;
+        lock (this.syncObject)
+        {
+            if (this.value == DisposedSentinel)
+            {
+                return Task.CompletedTask;
+            }
+
+            switch (this.value?.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    // We'll dispose of the value inline, outside the lock.
+                    localValue = this.value.Result as IDisposable;
+                    break;
+                case TaskStatus.Faulted:
+                case TaskStatus.Canceled:
+                    // Nothing left to do.
+                    break;
+                default:
+                    // We'll schedule the value for disposal outside the lock so it can be synchronous with the value factory,
+                    // but will not execute within our lock.
+                    localValueTask = this.value;
+                    break;
+            }
+
+            // Shut out all future callers from obtaining the value.
+            this.value = DisposedSentinel;
+
+            // We want value to be set before valueFactory is cleared so that IsValueCreated never returns true incorrectly.
+            Interlocked.MemoryBarrier();
+
+            // Release associated memory.
+            this.joinableTask = null;
+            this.valueFactory = null;
+        }
+
+        if (localValue is not null)
+        {
+            localValue.Dispose();
+        }
+        else if (localValueTask is not null)
+        {
+            return Task.Run(async delegate
```

Comment:
It would not typically run inline here regardless, because this path is for value factories that are _in progress_, where it's impossible to dispose the value inline. 
That said, I think I was over-eager to optimize here. I was trying to avoid allocating a Task where it isn't necessary, but an `async Task` method (one with no T result) doesn't allocate, so I can take the easier path that you recommend without any adverse effect. I'll make the change.